### ISSUE TITLE
Terminal mode: The cursor stays a solid block even when the terminal window is unfocused

### ISSUE THREAD

**Issue by mortie** [neovim/neovim#14320]
In all other modes with a block cursor, the cursor is a solid block when the window is focused, and becomes an outline of the block when the window is unfocused. In terminal mode, the cursor is a solid block even when the window is unfocused. This is a problem because it causes me to frequently input text into the wrong window.

- `nvim --version`: NVIM v0.5.0-dev+1231-g48e805728
- `vim -u DEFAULTS` (version: 8.2) behaves differently? Yes; the cursor in vim's `:terminal` works as expected
- Operating system/version: Ubuntu 21.04
- Terminal name/version: `GNOME Terminal 3.38.1 using VTE 0.62.3 +BIDI +GNUTLS +ICU +SYSTEMD` (also reproducible in kitty, xterm and urxvt)
- `$TERM`: xterm-256color

### Steps to reproduce using `nvim -u NORC`

1. Verify that the block cursor is solid when the terminal window is focused, and outlined when the terminal window is not focused.
2. Create a terminal with `:terminal`
3. Enter TERMINAL mode with `i`
4. Unfocus the terminal window (by focusing another window)

### Actual behaviour

The cursor stays a solid block when the window is unfocused.

### Expected behaviour

The cursor should become an outline of a block when the window is unfocused.

Here's an image which shows the issue: 
![Bad Cursor](https://user-images.githubusercontent.com/3728194/114006842-15355b00-9861-11eb-89f3-c081d3e5d9e3.png)

**Comment by glacambre** [comment]
I think this is normal, in terminal mode I expect the application running in the terminal to be the one who decides what the cursor looks like and neovim shouldn't have a say IMO.

If you go back to normal mode in the terminal buffer you'll see that the cursor correctly changes shape. A surer way to determine whether neovim has focus would be to change your colorscheme on FocusLost/FocusGained autocommands.

**Comment by mortie** [comment]
I don't understand why neovim's terminal mode should work differently from vim's terminal mode, tmux, or the plain terminal in this respect? If the application in the terminal asks to change the cursor then sure, respect that, but in the absence of that, wouldn't it be appropriate to just have a sensible default behavior?

**Comment by glacambre** [comment]
>I don't understand why neovim's terminal mode should work differently from vim's terminal mode, tmux, or the plain terminal in this respect?

Because Neovim is none of these programs? ;)

>If the application in the terminal asks to change the cursor then sure, respect that, but in the absence of that, wouldn't it be appropriate to just have a sensible default behavior?

Yeah, I agree, in the case where the application being executed didn't specify a cursor shape, it'd make sense for neovim to hollow it out as it already allows itself to change the cursor's look when changing windows. I see this more as a feature request than a bug though :).

**Comment by bew** [comment]
The block-outline-cursor isn't something neovim sets or _can_ set afaik, this is something the GNOME Terminal should do whenever the window is selected or not.
Neovim in this case is not really aware whether it is focused or not.
Right? This definitely does not explain what you see.
Maybe it is because Neovim sets the cursor shape to a special block when in terminal mode (vs normal mode), that GNOME Terminal does not change (due to a big?) to an outline when unfocused?

**Comment by mortie** [comment]
I want to clarify that this is _definitely_ not a gnome-terminal bug, because the same behavior happens in kitty, urxvt and even xterm (and probably many, many others; those are just the ones I tested). If it's a terminal bug, then it's a bug that's so wide-spread by now that it's basically a feature.

**Comment by erw7** [comment]
The change in the shape of the cursors due to the focus state of the terminal is a function of the terminal (GNOME terminal etc). However, the cursor displayed on the neovim terminal is not the cursor of the host terminal, but is drawn by neovim by highlighting. Therefore, this is a neovim issue.

For now, how about the following settings as a workaround?

```vim
autocmd FocusLost * hi TermCursor cterm=NONE gui=NONE
autocmd FocusGained * hi TermCursor cterm=reverse gui=reverse
```

**Comment by glacambre** [comment]
Since the solution to this issue is to stop drawing a cursor and to use the terminal's cursor instead, this issue can probably be considered a duplicate of https://github.com/neovim/neovim/issues/3681 .

**Comment by SampoLS** [comment]
Please tell me how to set the outline of the block cursor!!!

**Comment by zeertzjq** [comment]
> Please tell me how to set the outline of the block cursor in neovim!!!

What do you mean by "outline"?

**Comment by zeertzjq** [comment]
> Please tell me how to set the outline of the block cursor in neovim!!!

It has nothing to do with Neovim. Cursor shape is a feature of terminal emulators.

**Comment by zeertzjq** [comment]
> Please tell me how to set the outline of the block cursor in neovim!!!

It has nothing do with Neovim. That is done by the terminal emulator.

**Comment by zeertzjq** [comment]
> Please tell me how to set the outline of the block cursor in neovim!!!

You question is not very clear. Are you asking about in `:terminal` Terminal mode, or in other modes?

In `:terminal` Terminal mode, this is currently impossible.

In others modes, it has nothing to do with Nvim. Your terminal emulator needs to have that feature, and not every terminal emulator has that feature.

**Comment by SampoLS** [comment]
> > Please tell me how to set the outline of the block cursor in neovim!!!
> 
> You question is not very clear. Are you asking about in `:terminal` Terminal mode, or in other modes?
> 
> In `:terminal` Terminal mode, this is currently impossible.
> 
> In others modes, it has nothing to do with Nvim. Your terminal emulator needs to have that feature, and not every terminal emulator has that feature.

Thanks for your replying. Yeah, I installed the neovim extension in vscode and I wanna change the cursor shape. So, I changed the cursor to block shape in _insert mode_ by setting **set guicursor=i:block** in init.vim and I made it in both vscode and windows terminal. But I wanna change the cursor shape to outline-block in **normal mode**, I have no idea about how to configure it in init.vim file.  I searched a lot in google but still confused on how to fix it, this is where I got stuck . All right, my system is windows 10 and I'm a newbie in neovim. Could you help me?

**Comment by zeertzjq** [comment]
> But I wanna change the cursor shape to outline-block in **normal mode**, I have no idea about how to configure it in init.vim file.

You question is off-topic and I don't know much about it. IMO there isn't such a feature currently. I suggest you open a feature request.

**Comment by zeertzjq** [comment]
I think this can be closed as a duplicate of #3681
